File: packages/create-llama/templates/components/vectordbs/llamaindexserver/llamacloud/typescript/generate.ts
import { generateEventComponent } from "@llamaindex/server";
import * as dotenv from "dotenv";
import "dotenv/config";
import * as fs from "fs/promises";
import { LLamaCloudFileService, OpenAI } from "llamaindex";
import * as path from "path";
import { getIndex } from "./app/data";
import { initSettings } from "./app/settings";

dotenv.config();

const REQUIRED_ENV_VARS = [
  "LLAMA_CLOUD_INDEX_NAME",
  "LLAMA_CLOUD_PROJECT_NAME",
  "LLAMA_CLOUD_API_KEY",
];

export function checkRequiredEnvVars() {
  const missingEnvVars = REQUIRED_ENV_VARS.filter((envVar) => {
    return !process.env[envVar];
  });

  if (missingEnvVars.length > 0) {
    console.log(
      `The following environment variables are required but missing: ${missingEnvVars.join(
        ", ",
      )}`,
    );
    throw new Error(
      `Missing environment variables: ${missingEnvVars.join(", ")}`,
    );
  }
}

async function* walk(dir: string): AsyncGenerator<string> {
  const directory = await fs.opendir(dir);

  for await (const dirent of directory) {
    const entryPath = path.join(dir, dirent.name);

    if (dirent.isDirectory()) {
      yield* walk(entryPath); // Recursively walk through directories
    } else if (dirent.isFile()) {
      yield entryPath; // Yield file paths
    }
  }
}

async function loadAndIndex() {
  const index = await getIndex();
  // ensure the index is available or create a new one
  await index.ensureIndex({
    verbose: true,
    embedding: {
      type: "OPENAI_EMBEDDING",
      component: {
        api_key: process.env.OPENAI_API_KEY,
        model_name: "text-embedding-3-small",
      },
    },
  });

  const projectId = await index.getProjectId();
  const pipelineId = await index.getPipelineId();

  // walk through the data directory and upload each file to LlamaCloud
  for await (const filePath of walk("data")) {
    const buffer = await fs.readFile(filePath);
    const filename = path.basename(filePath);
    try {
      await LLamaCloudFileService.addFileToPipeline(
        projectId,
        pipelineId,
        new File([buffer], filename),
      );
    } catch (error) {
      if (
        error instanceof ReferenceError &&
        error.message.includes("File is not defined")
      ) {
        throw new Error(
          "File class is not supported in the current Node.js version. Please use Node.js 20 or higher.",
        );
      }
      throw error;
    }
  }

  console.log(`Successfully uploaded documents to LlamaCloud!`);
}

async function generateDatasource() {
  try {
    checkRequiredEnvVars();
    initSettings();
    await loadAndIndex();
    console.log("Finished generating storage.");
  } catch (error) {
    console.error("Error generating storage.", error);
  }
}

async function generateUi() {
  // Also works well with Claude 3.5 Sonnet and Google Gemini 2.5 Pro
  const llm = new OpenAI({ model: "gpt-4.1" });

  const workflowModule = await import("./app/workflow");
  const UIEventSchema = (workflowModule as any).UIEventSchema;
  if (!UIEventSchema) {
    throw new Error(
      "To generate the UI, you must define a UIEventSchema in your workflow.",
    );
  }

  const generatedCode = await generateEventComponent(UIEventSchema, llm);
  // Write the generated code to components/ui_event.ts
  await fs.writeFile("components/ui_event.jsx", generatedCode);
}

(async () => {
  const args = process.argv.slice(2);
  const command = args[0];

  initSettings();

  if (command === "datasource") {
    await generateDatasource();
  } else if (command === "ui") {
    await generateUi();
  } else {
    console.error(
      'Invalid command. Please use "datasource" or "ui". Running "datasource" by default.',
    );
    await generateDatasource(); // Default behavior or could throw an error
  }
})();

File: packages/create-llama/templates/components/vectordbs/llamaindexserver/llamacloud/python/generate.py
# flake8: noqa: E402

from dotenv import load_dotenv

load_dotenv()

import logging

from llama_index.core.readers import SimpleDirectoryReader
from tqdm import tqdm

from src.index import get_index
from src.service import LLamaCloudFileService
from src.settings import init_settings

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger()


def generate_index():
    init_settings()
    logger.info("Generate index for the provided data")

    index = get_index(create_if_missing=True)
    if index is None:
        raise ValueError("Index not found and could not be created")

    # use SimpleDirectoryReader to retrieve the files to process
    reader = SimpleDirectoryReader(
        "ui/data",
        recursive=True,
    )
    files_to_process = reader.input_files

    # add each file to the LlamaCloud pipeline
    error_files = []
    for input_file in tqdm(
        files_to_process,
        desc="Processing files",
        unit="file",
    ):
        with open(input_file, "rb") as f:
            logger.debug(
                f"Adding file {input_file} to pipeline {index.name} in project {index.project_name}"
            )
            try:
                LLamaCloudFileService.add_file_to_pipeline(
                    index.project.id,
                    index.pipeline.id,
                    f,
                    custom_metadata={},
                    wait_for_processing=False,
                )
            except Exception as e:
                error_files.append(input_file)
                logger.error(f"Error adding file {input_file}: {e}")

    if error_files:
        logger.error(f"Failed to add the following files: {error_files}")

    logger.info("Finished generating the index")


if __name__ == "__main__":
    generate_index()
